### PR TITLE
DBZ-177: Providing an alternative way to create JDBC connection based …

### PR DESCRIPTION
**Pull request by rareddy**
…on the configured JDBC driver class name and supplied classloader. The loading/creating the JDBC connections is not reliable when driver libraries in a different classloader than the DriverManager. This also preserves previous behavior intact.

### CODE REVIEW

File: debezium-core/src/main/java/io/debezium/jdbc/JdbcConnection.java [hunk 1]
```diff
@@ -97,7 +97,7 @@
      * @param variables any custom or overridden configuration variables
      * @return the connection factory
      */
-    public static ConnectionFactory patternBasedFactory(String urlPattern, Field... variables) {
+    public static ConnectionFactory patternBasedFactory(String urlPattern, ClassLoader classloader, Field... variables) {
```
Comment (rhauch) [inline debezium-core/src/main/java/io/debezium/jdbc/JdbcConnection.java]: Correct the JavaDoc, and specify that `classloader` may be null (see below). Also, we need to keep the old version of this method for backward compatibility, but it can simply delegate to this new signature.

File: debezium-core/src/main/java/io/debezium/jdbc/JdbcConnection.java [hunk 2]
```diff
@@ -110,7 +110,21 @@ public static ConnectionFactory patternBasedFactory(String urlPattern, Field...
...
-            Connection conn = DriverManager.getConnection(url, props);
+            
+            Connection conn = null;
+            String driverClassName = config.getString(JdbcConfiguration.DRIVER);
+            if (driverClassName == null) {
+                conn = DriverManager.getConnection(url, props);
+            } else {
+                try {
+                    @SuppressWarnings("unchecked")
+                    Class<java.sql.Driver> driverClazz = (Class<java.sql.Driver>) Class.forName(driverClassName, true, classloader);
```
Comment (rhauch) [inline debezium-core/src/main/java/io/debezium/jdbc/JdbcConnection.java]: What happens if `classloader` is null? I think `Class.forName(...)` will generate an exception, and if so you'll have to add an if-block to handle this case.

Comment (rareddy) [inline debezium-core/src/main/java/io/debezium/jdbc/JdbcConnection.java]: hmm, did not think about that. May be I can just supply the default JDBCConnection's classloader as default.

**Comment by rareddy** [comment]
Thanks, corrected the issues in next commit. Thanks for reviewing.

File: debezium-core/src/main/java/io/debezium/jdbc/JdbcConnection.java [hunk 3]
```diff
@@ -98,6 +98,27 @@
...
+     * <li><code>${port}</code></li>
+     * <li><code>${dbname}</code></li>
+     * <li><code>${username}</code></li>
+     * <li><code>${password}</code></li>
+     * </ul>
+     * 
+     * @param urlPattern the URL pattern string; may not be null
+     * @param variables any custom or overridden configuration variables
+     * @param classloader when configuration has jdbc.driver specified, the driver will be loaded from this 
+     *        class loader if one is provided. When null DriverManager will resolve the driver based on urlPattern
```
Comment (rhauch) [inline debezium-core/src/main/java/io/debezium/jdbc/JdbcConnection.java]: This isn't correct, is it? When the _driver class name_ is null, the DriverManager will resolve the driver using the URL patter. Otherwise, the this method will attempt to load the driver class (optionally using the `classloader` if one is provided), create a new instance of the Driver with the URL and properties, and then attempt to connect.

The JavaDoc for the method needs to explain what this is doing.

File: debezium-core/src/main/java/io/debezium/jdbc/JdbcConnection.java [hunk 4]
```diff
@@ -110,7 +131,25 @@ public static ConnectionFactory patternBasedFactory(String urlPattern, Field...
...
+            if (driverClassName == null) {
+                conn = DriverManager.getConnection(url, props);
+            } else {
+                try {
+                    Class<java.sql.Driver> driverClazz = null;
+                    if (classloader == null) {
+                        driverClazz = (Class<java.sql.Driver>) Class.forName(driverClassName);
+                    } else {
+                        driverClazz = (Class<java.sql.Driver>) Class.forName(driverClassName, true, classloader);
+                    }
```
Comment (rhauch) [inline debezium-core/src/main/java/io/debezium/jdbc/JdbcConnection.java]: This probably isn't as consistent with `patternBasedFactory(String, Field...)`. Seems like that method could pass null into the `classloader` parameter, and then this method should set `classloader = JdbcConnection.class.getClassLoader()` when `classloader` is null.

**Comment by rareddy** [comment]
I took your javadoc description verbatim :) Corrected the other.

File: debezium-core/src/main/java/io/debezium/jdbc/JdbcConnection.java [hunk 5]
```diff
@@ -98,6 +98,28 @@
...
+     * </ul>
+     * 
+     * @param urlPattern the URL pattern string; may not be null
+     * @param variables any custom or overridden configuration variables
+     * @param classloader When the driver class name is null, the DriverManager will resolve the driver using the URL pattern. 
+     *        Otherwise, the this method will attempt to load the driver class (optionally using the classloader if one is 
+     *        provided), create a new instance of the Driver with the URL and properties, and then attempt to connect. 
+     * @return the connection factory
+     */
+    @SuppressWarnings("unchecked")
```
Comment (rhauch) [inline debezium-core/src/main/java/io/debezium/jdbc/JdbcConnection.java]: I still don't think the JavaDoc really does a good job explaining what this method does, since you have to read the `classloader` parameter description to really know what's going to happen. And the order of the parameters still is not right. Please replace the whole JavaDoc for this method with the following:

```
    /**
     * Create a {@link ConnectionFactory} that uses the specific JDBC driver class loaded with the given class loader, and obtains the connection URL by replacing the following variables in the URL pattern:
     * <ul>
     * <li><code>${hostname}</code></li>
     * <li><code>${port}</code></li>
     * <li><code>${dbname}</code></li>
     * <li><code>${username}</code></li>
     * <li><code>${password}</code></li>
     * </ul>
     * <p>
     * This method attempts to instantiate the JDBC driver class and use that instance to connect to the database.
     * @param urlPattern the URL pattern string; may not be null
     * @param driverClassName the name of the JDBC driver class; may not be null
     * @param classloader the ClassLoader that should be used to load the JDBC driver class given by `driverClassName`; may be null if this class' class loader should be used
     * @param variables any custom or overridden configuration variables
     * @return the connection factory
     */
```

File: debezium-core/src/main/java/io/debezium/jdbc/JdbcConnection.java [hunk 6]
```diff
@@ -110,7 +132,24 @@ public static ConnectionFactory patternBasedFactory(String urlPattern, Field...
...
+                    if (driverClassLoader == null) {
+                        driverClassLoader = JdbcConnection.class.getClassLoader();
+                    }
+                    Class<java.sql.Driver> driverClazz = (Class<java.sql.Driver>) Class.forName(driverClassName, true, driverClassLoader);
+                    java.sql.Driver driver = driverClazz.newInstance();
+                    conn = driver.connect(url, props);
+                } catch (ClassNotFoundException|IllegalAccessException|InstantiationException e) {
+                    throw new SQLException(e);
+                }                
+            }
```
Comment (rhauch) [inline debezium-core/src/main/java/io/debezium/jdbc/JdbcConnection.java]: This whole method is relatively complex now. It really would be simpler if the old `patternBasedFactory(String urlPattern, Field...fields)` method did what it used to do (using `DriverManager.getConnection(url, props)`) and this method always used the `driverClassName` provided as a parameter and created an instance. Yes, it duplicates a bit of code, but each method is much more direct/obvious and there are fewer possible paths through each method.

Comment (rareddy) [inline debezium-core/src/main/java/io/debezium/jdbc/JdbcConnection.java]: Yes, agree. We will close this one. I will submit another pull request with those changes. Also, how do you expect my code to call the new pattern based factory? You want me to extend the "MySqlJdbcContext"?

Comment (rhauch) [inline debezium-core/src/main/java/io/debezium/jdbc/JdbcConnection.java]: No, it's not necessary to close this PR. Simply revert the changes and squash your comments into one (that's fine on a PR that has not been merged).

Why would you need to extend `MySqlJdbcContext`?

Comment (rhauch) [inline debezium-core/src/main/java/io/debezium/jdbc/JdbcConnection.java]: Do you *need* to pass in the name of the driver class? Seems like you want to provide your own classloader, so if you're doing that passing in a specific name of a driver class isn't that much harder. But maybe I don't understand the mechanics of how you're configuring and initializing the connector.

**Comment by rareddy** [comment]
The classloader is provided through the EmbeddedEngine.using method. Since this classloader is loads the connector classes, MySqlJdbcContext classloader is sent as classloader to the ConnctionFactory.

The driver name is sent through the configuration. This is also helpful in future if we want switch with different driver.

Ignore the extending the MySqlJdbcContext comment, I did not look code close enough before. I submitted another patch squashing all the commits into one.

**Comment by rareddy** [comment]
@rhauch Added the configuration constant to the previous commit and updated the pull. Pls take look.

File: debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/MySqlConnectorConfig.java [hunk 7]
```diff
@@ -410,6 +410,14 @@ public static SecureConnectionMode parse(String value, String defaultValue) {
                                                            .withDependents(DATABASE_WHITELIST_NAME)
                                                            .withDescription("Flag specifying whether built-in tables should be ignored.");
 
+    public static final Field JDBC_DRIVER = Field.create("database.jdbc.driver")
+            .withDisplayName("Jdbc Driver Class Name")
+            .withType(Type.STRING)
+            .withWidth(Width.MEDIUM)
+            .withDefault("com.mysql.jdbc.Driver")
```
Comment (rhauch) [inline debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/MySqlConnectorConfig.java]: Would it be better to define the default value as `com..mysq.jdbc.Driver.class.getName()` rather than a literal string?

File: debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/MySqlConnectorConfig.java [hunk 8]
```diff
@@ -410,6 +410,14 @@ public static SecureConnectionMode parse(String value, String defaultValue) {
                                                            .withDependents(DATABASE_WHITELIST_NAME)
                                                            .withDescription("Flag specifying whether built-in tables should be ignored.");
 
+    public static final Field JDBC_DRIVER = Field.create("database.jdbc.driver")
+            .withDisplayName("Jdbc Driver Class Name")
+            .withType(Type.STRING)
```
Comment (rhauch) [inline debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/MySqlConnectorConfig.java]: Also, I think it'd be better to use `Type.CLASS` here rather than `Type.STRING`, since the former basically specifies a literal string that is a class name.

Comment (rareddy) [inline debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/MySqlConnectorConfig.java]: oops, forgot, will fix shortly..

Comment (rareddy) [inline debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/MySqlConnectorConfig.java]: Since that will get you back a Class instead of string, I thought we wanted to keep it as string and validate as class?

Comment (rhauch) [inline debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/MySqlConnectorConfig.java]: That's what `Type.CLASS` does ... it simply means `Type.STRING` except the Confluent UI knows/treats it as a Java class name. It's a minimal validation when the UI user is entering values, and we still do proper validation via the `.withValidation(Field::isClassName)` call a few lines below this one.

Comment (rhauch) [inline debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/MySqlConnectorConfig.java]: Oh, maybe what you're referring to is that that the `Configuration` class does have methods to return a new instance of the class specified in a `Field`, but that's merely optional. You can always use `Configuration.getString(Field)` to get the string version of the property (field) value.

Comment (rareddy) [inline debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/MySqlConnectorConfig.java]: I am running into a test fail with that CLASS and `Validation::isClassName`
{code}
  MySqlConnectorTest.shouldReturnConfigurationDefinition:22->assertConfigDefIsValid:28->lambda$assertConfigDefIsValid$0:40 expected:<['com.mysql.jdbc.Driver']> but was:<[com.mysql.jdbc.Driver]>
{code}
That means I can't supply a default c`om.mysql.jdbc.Driver.class.getName()` any more. If I use the  `com.mysql.jdbc.Driver.class` as default, then `withDefault` method is not designed for class types. Also note, once I do this, then the configuration is going to return a <class> definition not string form of class right? If that is case then this will force me to further changes on look out for class itself not name and classloader to resolve in further downstream code?

Comment (rhauch) [inline debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/MySqlConnectorConfig.java]: Like [this Field definition](https://github.com/debezium/debezium/blob/master/debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/MySqlConnectorConfig.java#L570)? We use that field all the time.

Comment (rareddy) [inline debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/MySqlConnectorConfig.java]: @rhauch  Hmm, then I do not understand why the unit test is failing?

Comment (rareddy) [inline debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/MySqlConnectorConfig.java]: Ok, I see that that there is exception clause to handle this in the unit test. Will do the same then.

**Comment by rareddy** [comment]
@rhauch corrected the testcase and modified the to CLASS type now.

**Comment by rhauch** [comment]
Looks great, @rareddy - thanks!
